src/api-service/__app__/onefuzzlib/agent_events.py

```diff
@@ -0,0 +1,239 @@
+        # if tasks are running on the node when it reports as Done
+        # those are stopped early
+        node.mark_tasks_stopped_early()
+
+        # Model-validated.
+        #
+        # This field will be required in the future.
+        # For now, it is optional for back compat.
+        done_data = cast(Optional[NodeDoneEventData], state_update.data)
+        if done_data:
+            # TODO: do something with this done data
+            if done_data.error:
+                logging.error(
+                    "node 'done' with error: machine_id:%s, data:%s",
+                    machine_id,
+                    done_data,
+                )
+    return None
+
+
+def on_worker_event_running(
+    machine_id: UUID, event: WorkerRunningEvent
+) -> Optional[Error]:
+    task = Task.get_by_task_id(event.task_id)
+    if isinstance(task, Error):
+        return task
+
+    node = get_node(machine_id)
+    if isinstance(node, Error):
+        return node
+
+    node_task = NodeTasks(
+        machine_id=machine_id, task_id=event.task_id, state=NodeTaskState.running
+    )
+    node_task.save()
+
+    if task.state in TaskState.shutting_down():
+        logging.info(
+            "ignoring task start from node.  machine_id:%s %s:%s (state: %s)",
+            machine_id,
+            task.job_id,
+            task.task_id,
+            task.state,
+        )
+        return None
+
+    logging.info(
+        "task started on node.  machine_id:%s %s:%s",
+        machine_id,
+        task.job_id,
+        task.task_id,
+    )
+    task.state = TaskState.running
+    task.save()
+
+    # Start the clock for the task if it wasn't started already
+    # (as happens in 1.0.0 agents)
+    task.on_start()
+
+    return None
+
+
+def on_worker_event_done(machine_id: UUID, event: WorkerDoneEvent) -> Optional[Error]:
+    task = Task.get_by_task_id(event.task_id)
+    if isinstance(task, Error):
+        return task
+
+    node = get_node(machine_id)
+    if isinstance(node, Error):
+        return node
+
+    node_task = NodeTasks(
+        machine_id=machine_id, task_id=event.task_id, state=NodeTaskState.running
+    )
+    node_task.delete()
+
+    exit_status = event.exit_status
+    if not exit_status.success:
+        logging.error(
+            "task failed. %s:%s status:%s", task.job_id, task.task_id, exit_status
+        )
+        task.mark_failed(
+            Error(
+                code=ErrorCode.TASK_FAILED,
+                errors=[
+                    "task failed. exit_status:%s" % exit_status,
+                    event.stdout[-4096:],
+                    event.stderr[-4096:],
+                ],
+            )
+        )
+
+        if task.config.debug and (
+            TaskDebugFlag.keep_node_on_failure in task.config.debug
+            or TaskDebugFlag.keep_node_on_completion in task.config.debug
+        ):
+            node.debug_keep_node = True
+            node.save()
+
+    else:
+        logging.error(
```

Comment:
This shouldn't be error; we're in the `not not exit_status.success` branch.

Maybe we should condition on `if exit_status.success:`, instead.

Reply:
I swapped the check around, and moved that to 'info'